### PR TITLE
Miscellaneous Builder changes

### PR DESCRIPTION
**Pull request by move47**
- Validates builder message signatures.
- Add the builder fee signatures.
- Removes the spawn blocking over vid calls. 
- Changes the first API call to be based on builder commitment instead of VID commitment.

### CODE REVIEW

File: crates/task-impls/src/helpers.rs [hunk 1]
```diff
@@ -60,6 +60,29 @@ pub async fn calculate_vid_disperse<TYPES: NodeType>(
...
+/// # Panics
+/// Panics if the VID calculation fails, this should not happen.
+pub async fn calculate_vid_disperse_using_precompute_data<TYPES: NodeType>(
+    txns: Vec<u8>,
+    membership: Arc<TYPES::Membership>,
+    view: TYPES::Time,
+    pre_compute_data: VidPrecomputeData,
+) -> VidDisperse<TYPES> {
+    let num_nodes = membership.total_nodes();
+    let vid_disperse = spawn_blocking(move || {
```
Comment (jparr721) [inline crates/task-impls/src/helpers.rs:74]: Can we just move the #[allow(panic)] to the top of the function?

File: crates/task-impls/src/helpers.rs [hunk 2]
```diff
@@ -60,6 +60,29 @@ pub async fn calculate_vid_disperse<TYPES: NodeType>(
...
+        #[allow(clippy::panic)]
+        vid_scheme(num_nodes).disperse_precompute(&txns, &pre_compute_data).unwrap_or_else(|err|panic!("VID disperse failure:\n\t(num_storage nodes,payload_byte_len)=({num_nodes},{})\n\terror: : {err}", txns.len()))
```
Comment (jparr721) [inline crates/task-impls/src/helpers.rs:76]: Two things:
1. It looks like the formatter missed this, it seems like `|err|panic` should have a space.
2. I don't think it's a good idea to add new lines and tabs to error logs as it'll make it more difficult to parse when reading and applying structured logging.
3. You shouldn't need to panic here. Instead, you can remove `error: {err}` entirely and simply use `expect`.

Comment (jparr721) [reply crates/task-impls/src/helpers.rs:76]: As an addendum, I'm not crystal clear why we spawn this as a blocking task? Can't you just call this would needing a background thread at all?

File: crates/task-impls/src/transactions.rs [hunk 3]
```diff
@@ -144,16 +147,21 @@ impl<
     #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Transaction Handling Task", level = "error")]
     async fn wait_for_block(
         &self,
-    ) -> Option<(AvailableBlockData<TYPES>, AvailableBlockHeaderInput<TYPES>)> {
+    ) -> Option<(
```
Comment (jparr721) [inline crates/task-impls/src/transactions.rs:150]: This is becoming a bit hard to parse. We should maybe make this into a struct.

**Review by jparr721 — CHANGES_REQUESTED** [review]

File: crates/task-impls/src/helpers.rs (hunk 2 above)
Comment (nyospe) [reply crates/task-impls/src/helpers.rs:76]: I concur on the first to points. For the third... isn't `disperse_precompute` a bit long running, even with the precompute? I agree that `spawn_blocking` here is not helpful, but not on the reason. The advantage of `spawn` is that you can launch it allow it to run in a different thread, then come back and await (on the `JoinHandle`) after doing other stuff in parallel. It's a way to leverage the advantages of multithreaded concurrency in an async context. By contrast, `spawn_blocking` will yield on the calling async context while the spawned task gets to run (effectively) as if it were synchronous. This will allow other async tasks to progress, but not this current async task.

Crucially, either of these would only execute when you call `.await` on the future returned by `calculate_vid_disperse_using_precompute_data`, which isn't where the advantage of using `spawn` would be. To avoid serial execution of independent operations, you would want to `spawn` around `calculate_vid_disperse_using_precompute_data` and then await the result after completion of whatever other work needed doing in the calling context.

I suspect this is using `spawn_blocking` because of an (incorrect) rule of thumb that was widely promoted a few years ago, that recommended using it for longer (tens to hundreds of milliseconds) CPU-bound operations. This is still a one-off operation that eventually completes, and would be better handled with something like `rayon` or one of the more sophisticated threadpool crates.

File: crates/task-impls/src/helpers.rs (hunk 1 above)
Comment (move47) [reply crates/task-impls/src/helpers.rs:75]: https://github.com/EspressoSystems/HotShot/pull/2956/commits/8d15da890e00f3ab93edcd880b5fb1160b95768d

File: crates/task-impls/src/helpers.rs (hunk 2 above)
Comment (move47) [reply crates/task-impls/src/helpers.rs]: I guess it's fine since the existing api was following the same pattern. https://github.com/EspressoSystems/HotShot/blob/15c776d32001ce88edf2a87f3c520301da22a582/crates/task-impls/src/helpers.rs#L45

File: crates/task-impls/src/transactions.rs (hunk 3 above)
Comment (move47) [reply crates/task-impls/src/transactions.rs]: https://github.com/EspressoSystems/HotShot/pull/2956/commits/8d15da890e00f3ab93edcd880b5fb1160b95768d

File: crates/types/src/traits/block_contents.rs [hunk 4]
```diff
@@ -124,6 +124,21 @@ pub fn vid_commitment(
...
+#[must_use]
+pub fn precompute_vid_commitment(
+    encoded_transactions: &Vec<u8>,
+    num_storage_nodes: usize,
+) -> (
+    <VidSchemeType as VidScheme>::Commit,
+    <VidSchemeType as Precomputable>::PrecomputeData,
+) {
+    #[allow(clippy::panic)]
+     vid_scheme(num_storage_nodes).commit_only_precompute(encoded_transactions).unwrap_or_else(|err| panic!("VidScheme::commit_only failure:\n\t(num_storage_nodes,payload_byte_len)=({num_storage_nodes},{}\n\t{err}", encoded_transactions.len()))
```
[the remark this answers is not in the document]
Comment (move47) [reply crates/types/src/traits/block_contents.rs]: https://github.com/EspressoSystems/HotShot/pull/2956/commits/9eea1144928da906c5d7ba3dab4e4ac0be962111

**Comment by move47** [comment]
@jparr721 Can I get review again?

**Comment by jparr721** [comment]
> @jparr721 Can I get review again?

Looks like the error messages still have the newlines and tab characters, and it looks like we're still explicitly panicking. Can we remove both of those and I can re-remove?

**Comment by jparr721** [comment]
@nyospe Looks fine to me, but I am not deep on these changes. Mind taking a look as well?

File: crates/task-impls/src/transactions.rs [hunk 5]
```diff
@@ -29,6 +31,18 @@ use crate::{
     helpers::broadcast_event,
 };
 
+/// Builder Provided Responses
+pub struct BuilderResponses<TYPES: NodeType> {
```
Comment (nyospe) [inline crates/task-impls/src/transactions.rs:35]: I'm not clear on why we're bundling this all in one struct. Are we collecting what we get from each phase with the builder?

File: crates/task-impls/src/transactions.rs [hunk 6]
```diff
@@ -186,9 +199,32 @@ impl<
                 continue;
             };
 
+            // Verify signature over chosen block instead of
```
Comment (nyospe) [inline crates/task-impls/src/transactions.rs:202]: 👍

File: crates/task-impls/src/transactions.rs (hunk 5 above)
Comment (move47) [reply crates/task-impls/src/transactions.rs:35]: Yes, we are bundling all the responses from the builder. Later, we might filter and store what we need.

File: crates/builder-api/src/block_info.rs [hunk 7]
```diff
@@ -33,7 +33,12 @@ pub struct AvailableBlockData<TYPES: NodeType> {
 #[serde(bound = "")]
 pub struct AvailableBlockHeaderInput<TYPES: NodeType> {
     pub vid_commitment: VidCommitment,
-    pub signature:
+    pub vid_precompute_data: VidPrecomputeData,
```
Comment (nyospe) [inline crates/builder-api/src/block_info.rs:36]: This part looks good.

File: crates/task-impls/src/helpers.rs [hunk 8]
```diff
@@ -42,15 +42,38 @@ pub async fn broadcast_event<E: Clone + std::fmt::Debug>(event: E, sender: &Send
...
     view: TYPES::Time,
 ) -> VidDisperse<TYPES> {
     let num_nodes = membership.total_nodes();
     let vid_disperse = spawn_blocking(move || {
-        #[allow(clippy::panic)]
-        vid_scheme(num_nodes).disperse(&txns).unwrap_or_else(|err|panic!("VID disperse failure:\n\t(num_storage nodes,payload_byte_len)=({num_nodes},{})\n\terror: : {err}", txns.len()))
+        vid_scheme(num_nodes).disperse(&txns).unwrap_or_else(|err| panic!("VID disperse failure:(num_storage nodes,payload_byte_len)=({num_nodes},{}) error: {err}", txns.len()))
+    })
+    .await;
+    #[cfg(async_executor_impl = "tokio")]
```
Comment (nyospe) [inline crates/task-impls/src/helpers.rs:56]: I have questions. I don't think there's any other point in the HotShot repo where we hardcode only tokio. Maybe it would be worth unravelling this spawn_blocking just to avoid that.

File: crates/task-impls/src/helpers.rs [hunk 9]
```diff
@@ -42,15 +42,38 @@ pub async fn broadcast_event<E: Clone + std::fmt::Debug>(event: E, sender: &Send
...
+/// Panics if the VID calculation fails, this should not happen.
+#[allow(clippy::panic)]
+pub async fn calculate_vid_disperse_using_precompute_data<TYPES: NodeType>(
+    txns: Vec<u8>,
+    membership: Arc<TYPES::Membership>,
+    view: TYPES::Time,
+    pre_compute_data: VidPrecomputeData,
+) -> VidDisperse<TYPES> {
+    let num_nodes = membership.total_nodes();
+    let vid_disperse = spawn_blocking(move || {
```
Comment (nyospe) [inline crates/task-impls/src/helpers.rs:75]: This one too...

File: crates/task-impls/src/transactions.rs (hunk 5 above)
Comment (nyospe) [reply crates/task-impls/src/transactions.rs:35]: Doesn't that defeat the point of, well, having an external builder?

Comment (move47) [reply crates/task-impls/src/transactions.rs:35]: Sorry, I didn't get wdym?

File: crates/task-impls/src/helpers.rs (hunk 9 above)
Comment (move47) [reply crates/task-impls/src/helpers.rs]: https://github.com/EspressoSystems/HotShot/pull/2956/commits/5408f863bc99171c7e299ac0460875145d98252d

File: crates/task-impls/src/helpers.rs (hunk 8 above)
Comment (move47) [reply crates/task-impls/src/helpers.rs]: https://github.com/EspressoSystems/HotShot/pull/2956/commits/5408f863bc99171c7e299ac0460875145d98252d

File: crates/builder-api/src/data_source.rs [hunk 10]
```diff
@@ -15,7 +14,7 @@ pub trait BuilderDataSource<TYPES: NodeType> {
     /// To get the list of available blocks
     async fn get_available_blocks(
         &self,
-        for_parent: &VidCommitment,
+        for_parent: &BuilderCommitment,
```
Comment (nyospe) [inline crates/builder-api/src/data_source.rs:17]: 👍

File: crates/task-impls/src/transactions.rs (hunk 5 above)
Comment (nyospe) [reply crates/task-impls/src/transactions.rs:35]: Never mind, we can fix this during the transition to multi-builder auctions

**Review by nyospe — APPROVED** [review]
Okay, I think this is ready now.

**Review by jparr721 — APPROVED** [review]
